### DISCUSSION TITLE
How to use the CoreMutex class?

### DISCUSSION THREAD

**Discussion by joba-1** [earlephilhower/arduino-pico#1311] [Q&A] [answered]
I'd like to protect access to a data structure that is modified on core 1 and read by core 0.
I thought using class CoreMutex (as used by serial) is a good idea, but compile fails.
Is CoreMutex a class that is intended to be used by apps?
Shouldn't then CoreMutex.h include everything I need for the class?
Fixing the first error is simple by including sys/types.h, the second I still search a solution (other than implementing my own class).

```
Compiling .pio/build/picow_dap/src/radar.cpp.o
In file included from src/radar.h:6,
                 from src/radar.cpp:1:
/home/joachim/.platformio/packages/framework-arduinopico/cores/rp2040/CoreMutex.h:46:5: error: 'u_int8_t' does not name a type; did you mean 'uint8_t'?
   46 |     u_int8_t _option;
      |     ^~~~~~~~
      |     uint8_t
Compiling .pio/build/picow_dap/src/wifi.cpp.o
In file included from .pio/libdeps/picow_dap/ld2410/src/ld2410.h:15,
                 from src/radar.h:7,
                 from src/radar.cpp:1:
/home/joachim/.platformio/packages/framework-arduinopico/cores/rp2040/Arduino.h:88:13: error: conflicting declaration of 'bool __isFreeRTOS' with 'C' linkage
   88 | extern bool __isFreeRTOS;
      |             ^~~~~~~~~~~~
In file included from /home/joachim/.platformio/packages/framework-arduinopico/cores/rp2040/CoreMutex.h:27,
                 from src/radar.h:6,
                 from src/radar.cpp:1:
/home/joachim/.platformio/packages/framework-arduinopico/cores/rp2040/_freertos.h:28:13: note: previous declaration with 'C++' linkage
   28 | extern bool __isFreeRTOS;
      |             ^~~~~~~~~~~~

```

https://github.com/joba-1/PicoPresence/blob/master/src/radar.cpp#L4
https://github.com/joba-1/PicoPresence/blob/master/src/radar.h#L6

**Comment by earlephilhower** [comment, marked as answer]
Interesting observations!  `CoreMutex` could be used by and app, but it's so simple I didn't document it or try it outside of the core.

The `u_int8_t` is actually just a silly typo on my part, and was of course meant to be a `uint8_t` just like the constructor `option` parameter.

The `__isFreeRTOS` is just a C/C++ linkage issue where I seem to have declared it in `Arduino.h` inside an `extern "C"` block, but in the FreeRTOS shim layer header w/o the `extern "C"`.  The actual variable is `C++` linkage in `main.cpp` so Arduino.h needs to move it around.

Changes in #1312 should clear these two up.

**Reply by joba-1** [reply]
Works great. Thanks for the fast fix!

**Reply by joba-1** [reply]
> it's so simple I didn't document it

Well, should you at some point decide to document it, It would help to include that the mutex_t variable needs to be mutex_init()ed before using it with a CoreMutex. Guess how I found out ... yes, the hard way :)

**Reply by earlephilhower** [reply]
Yeah, mutexes here have some subtle gotchas but I think that's pretty standard.  Check out the `auto_init_mutex` macro which will init the mutex as part of the C runtime initialization.
